buffer/src/main/java/io/netty/buffer/api/CompositeBuffer.java

```diff
@@ -0,0 +1,2028 @@
+        }
+        if (srcPos + length > capacity) {
+            throw indexOutOfBounds(srcPos + length, false);
+        }
+        while (length > 0) {
+            var buf = (Buffer) chooseBuffer(srcPos, 0);
+            int toCopy = Math.min(buf.capacity() - subOffset, length);
+            dest.copyInto(buf, subOffset, destPos, toCopy);
+            srcPos += toCopy;
+            destPos += toCopy;
+            length -= toCopy;
+        }
+    }
+
+    @FunctionalInterface
+    private interface CopyInto {
+        void copyInto(Buffer src, int srcPos, int destPos, int length);
+    }
+
+    @Override
+    public void copyInto(int srcPos, Buffer dest, int destPos, int length) {
+        if (length < 0) {
+            throw new IndexOutOfBoundsException("Length cannot be negative: " + length + '.');
+        }
+        if (srcPos < 0) {
+            throw indexOutOfBounds(srcPos, false);
+        }
+        if (srcPos + length > capacity) {
+            throw indexOutOfBounds(srcPos + length, false);
+        }
+        if (dest.readOnly()) {
+            throw bufferIsReadOnly(dest);
+        }
+
+        // Iterate in reverse to account for src and dest buffer overlap.
+        // todo optimise by delegating to constituent buffers.
+        var cursor = openReverseCursor(srcPos + length - 1, length);
+        ByteOrder prevOrder = dest.order();
+        // We read longs in BE, in reverse, so they need to be flipped for writing.
+        dest.order(ByteOrder.LITTLE_ENDIAN);
+        try {
+            while (cursor.readLong()) {
+                length -= Long.BYTES;
+                dest.setLong(destPos + length, cursor.getLong());
+            }
+            while (cursor.readByte()) {
+                dest.setByte(destPos + --length, cursor.getByte());
+            }
+        } finally {
+            dest.order(prevOrder);
+        }
+    }
+
+    @Override
+    public ByteCursor openCursor() {
+        return openCursor(readerOffset(), readableBytes());
+    }
+
+    @Override
+    public ByteCursor openCursor(int fromOffset, int length) {
+        if (fromOffset < 0) {
+            throw new IllegalArgumentException("The fromOffset cannot be negative: " + fromOffset + '.');
+        }
+        if (length < 0) {
+            throw new IllegalArgumentException("The length cannot be negative: " + length + '.');
+        }
+        if (capacity < fromOffset + length) {
+            throw new IllegalArgumentException("The fromOffset+length is beyond the end of the buffer: " +
+                                               "fromOffset=" + fromOffset + ", length=" + length + '.');
+        }
+        int startBufferIndex = searchOffsets(fromOffset);
+        int off = fromOffset - offsets[startBufferIndex];
+        Buffer startBuf = bufs[startBufferIndex];
+        ByteCursor startCursor = startBuf.openCursor(off, Math.min(startBuf.capacity() - off, length));
+        return new ByteCursor() {
```

Comment:
ultranit: a bad habit some developers have is calling Mockito's `spy()` function on the return value of methods.   When this happens, it means that they can start making faulty tests that assume the state transitions.   As a result, we (gRPC) started making these anonymous classes final by making them named method classes.

As a benefit, the stack traces on exception in these classes, the className, and profiling tools have an easier time dealing with the method-local class names.

Consider naming these anonymous classes.

Reply:
I'll do that in another PR